src/iqm/qiskit_iqm/iqm_transpilation.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2023 Qiskit on IQM developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transpilation tool to optimize the decomposition of single-qubit gates tailored to IQM hardware."""
+
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.circuit.library import RGate
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes import Optimize1qGatesDecomposition, Unroller
+from qiskit.transpiler.passmanager import PassManager
+
+
+class IQMOptimize1QbDecomposition(TransformationPass):
+    r"""Optimize the decomposition of single-qubit gates for the IQM gate set.
+
+    This optimisation pass expects the circuit to be correctly layouted and translated to the IQM architecture
+    and raises an error otherwise.
+    The optimisation logic follows the steps:
+
+    1. Convert single-qubit gates to :math:`U` gates and combine all neighbouring :math:`U` gates.
+    2. Convert :math:`U` gates according to
+       :math:`U(\theta , \phi , \lambda) ~ RZ(\phi + \lambda) R(\theta, \pi / 2  - \lambda)`.
+    3. Commute `RZ` gates to the end of the circuit using the fact that `RZ` and `CZ` gates commute, and
+       :math:`R(\theta , \phi) RZ(\lambda) = RZ(\lambda) R(\theta, \phi - \lambda)`.
+    4. Drop `RZ` gates immediately before measurements, and otherwise replace them according to
+       :math:`RZ(\lambda) = R(\pi, \lambda / 2) R(- \pi, 0)`.
+    """
+
+    def __init__(self, drop_final_rz: bool = False):
```

Comment:
Can you explain the meaning of the optional argument in the above docstring. Add an Args section.

Reply:
Added